### PR TITLE
Add NFT example

### PR DESCRIPTION
**Pull request by LouiseMedova**
Resolves # .

-
-
-

@reviewer-or-team

### CODE REVIEW

File: nft-example/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,496 @@
+#![no_std]
+#![feature(const_btree_new)]
+
+use codec::{Decode, Encode};
+use gstd::{debug, exec, msg, prelude::*, ActorId};
+use primitive_types::{H256, U256};
+use scale_info::TypeInfo;
+
+pub mod payloads;
+<<<<<<< HEAD
```
Comment (vivekvpandya) [inline nft-example/src/lib.rs:10]: git rebase/merge is not completed properly. @LouiseMedova please check this.

Comment (LouiseMedova) [reply nft-example/src/lib.rs]: Thanks, fixed it

File: nft-example/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+    Transfer(TransferInput),
+    Approve(ApproveInput),
+    ApproveForAll(ApproveForAllInput),
+    OwnerOf(U256),
+    BalanceOf(H256),
+}
+
+#[derive(Debug, Encode, Decode, TypeInfo)]
+pub enum Event {
+    // Base(NftEvent),
```
Comment (vivekvpandya) [inline nft-example/src/lib.rs:34]: is this required?

**Comment by vivekvpandya** [comment]
I think https://github.com/gear-tech/apps/pull/9/files is subsumed in this PR, so we can close that.

**Comment by vivekvpandya** [comment]
Please add comments (describing intention of message and expected outcome in log part) in test.yaml as it becomes hard to follow long tests. I will also update my FungibleToken test file similarly :)

File: nft-example/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+use primitive_types::{H256, U256};
+use scale_info::TypeInfo;
+
+pub mod payloads;
+pub use payloads::{ApproveForAllInput, ApproveInput, InitConfig, TransferInput};
+
+pub mod state;
+pub use state::{State, StateReply};
+
+use non_fungible_token::base::NonFungibleTokenBase;
```
Comment (dkubatko) [inline nft-example/src/lib.rs:15]: This import is redundant, please remove.

**Review by dkubatko — CHANGES_REQUESTED** [review]

File: nft-example/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+        }
+        Action::ApproveForAll(input) => {
+            CONTRACT.tokens.approve_for_all(
+                &msg::source(),
+                &ActorId::new(input.to.to_fixed_bytes()),
+                input.approve,
+            );
+        }
+        Action::OwnerOf(input) => {
+            let owner = CONTRACT.tokens.owner_by_id.get(&input).unwrap_or(&ZERO_ID);
```
Comment (dkubatko) [inline nft-example/src/lib.rs:165]: When the interface is updated, please replace this with `CONTRACT.tokens.owner_of` or similar.

File: nft-example/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+            msg::reply(
+                Event::OwnerOf(H256::from_slice(owner.as_ref())),
+                exec::gas_available() - GAS_RESERVE,
+                0,
+            );
+        }
+        Action::BalanceOf(input) => {
+            let balance = *CONTRACT
```
Comment (dkubatko) [inline nft-example/src/lib.rs:173]: Same as above.

File: nft-example/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+
+#[derive(Debug)]
+pub struct NFT {
+    pub tokens: NonFungibleToken,
+    pub token_id: U256,
+    pub owner: ActorId,
+}
+
+static mut CONTRACT: NFT = NFT {
+    tokens: NonFungibleToken {
```
Comment (dkubatko) [inline nft-example/src/lib.rs:49]: Why is this named `tokens`? I would say `token`, or `token_interface` would make more sense.

**Review by dkubatko — CHANGES_REQUESTED** [review]

File: nft-example/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+use scale_info::TypeInfo;
+
+pub mod payloads;
+pub use payloads::{ApproveForAllInput, ApproveInput, InitConfig, TransferInput};
+
+pub mod state;
+pub use state::{State, StateReply};
+
+use non_fungible_token::base::NonFungibleTokenBase;
+use non_fungible_token::{Approve, ApproveForAll, NonFungibleToken, Transfer};
```
Comment (dkubatko) [inline nft-example/src/lib.rs:16]: Not sure why `Approve`, `ApproveForAll` and `Transfer` are imported here. Those are just used as action names below, am I wrong?

**Review by dkubatko — CHANGES_REQUESTED** [review]

File: nft-example/src/lib.rs (hunk 6 above)
Comment (LouiseMedova) [reply nft-example/src/lib.rs:49]: I thought about it and decided to name this variable in the plural, because it includes all minted tokens and information about them. But  probably it would be betted in single form

File: nft-example/src/lib.rs (hunk 3 above)
Comment (LouiseMedova) [reply nft-example/src/lib.rs:15]: It is necessary since ' items from traits can only be used if the trait is in scope'

File: nft-example/src/lib.rs (hunk 7 above)
Comment (LouiseMedova) [reply nft-example/src/lib.rs:16]: No, these are not the name of actions, these are the names of structs in enum `Events`. `Transfer` struct is necessary since it is used in `mint` and `burn`, but `Approve` and `ApproveForAll` are really redundant, thanks

Comment (LouiseMedova) [reply nft-example/src/lib.rs:16]: `Approve`, `ApproveForAll` are also necessary, these are structs of message replies, and must be defined in `gstd::metadata` (enum `Events`)

File: nft-example/src/lib.rs (hunk 4 above)
Comment (LouiseMedova) [reply nft-example/src/lib.rs]: Updated interface and example, please check

File: nft-example/src/lib.rs (hunk 5 above)
Comment (LouiseMedova) [reply nft-example/src/lib.rs]: Updated interface and example, please check

File: nft-example/src/lib.rs (hunk 3 above)
Comment (vivekvpandya) [reply nft-example/src/lib.rs:15]: `make pre-commit` errors if any unused imports found.

File: nft-example/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+        balances: BTreeMap::new(),
+        operator_approval: BTreeMap::new(),
+    },
+    token_id: U256::zero(),
+    owner: ActorId::new(H256::zero().to_fixed_bytes()),
+};
+
+impl NFT {
+    fn mint(&mut self) {
+        self.token.owner_by_id.insert(self.token_id, msg::source());
```
Comment (vivekvpandya) [inline nft-example/src/lib.rs:65]: we may want to limit users who have privilege for minting new tokens.

File: nft-example/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+    OwnerOf(U256),
+    BalanceOf(H256),
+}
+
+#[derive(Debug, Encode, Decode, TypeInfo)]
+pub enum Event {
+    Transfer(Transfer),
+    Approval(Approve),
+    ApprovalForAll(ApproveForAll),
+    OwnerOf(H256),
```
Comment (vivekvpandya) [inline nft-example/src/lib.rs:37]: Suggestion: May be user `ActorId` instead of `H256`. But can be done in separate PR

Comment (LouiseMedova) [reply nft-example/src/lib.rs:37]: I will replace it in that PR, thank you

File: nft-example/src/lib.rs (hunk 8 above)
Comment (LouiseMedova) [reply nft-example/src/lib.rs:65]: There may be different use cases. Limiting users can be easily added, if it's required

File: gear/gtest/spec/test_nft_example.yaml [hunk 10]
```diff
@@ -0,0 +1,572 @@
...
+            destination: 2
+            payload:
+              kind: custom
+              value:
+                transfer:
+                  from: &zero 0x0000000000000000000000000000000000000000000000000000000000000000
+                  to: &user_1 0x0200000000000000000000000000000000000000000000000000000000000000
+                  token_id: 0 
+          # successfull reply of mint              
+          - <<: *mint_success
```
Comment (vivekvpandya) [inline gear/gtest/spec/test_nft_example.yaml:63]: I don't understand why `<<: *mint_success` is used here, is intention is to change value of `token_id: 1` in previous definition of `mint_success`?

Comment (LouiseMedova) [reply gear/gtest/spec/test_nft_example.yaml:63]: Yes

**Review by dkubatko — APPROVED** [review]

File: gear/gtest/spec/test_nft_example.yaml (hunk 10 above)
Comment (vivekvpandya) [reply gear/gtest/spec/test_nft_example.yaml:63]: but then it almost writing whole `mint_success` but I hope it is being used later.

Comment (LouiseMedova) [reply gear/gtest/spec/test_nft_example.yaml:63]: I just wanted to indicate that it's the same type of message (about successful mint) :)

**Review by shamilsan — APPROVED** [review]

**Review by vivekvpandya — APPROVED** [review]
